Modify per-version request handler support to use ClassInfo, not handler function

gavin-norman-sociomantic opened the pull request:
The impetus for this change is to fix the issue discovered in
sociomantic-tsunami/dhtproto#93:
* The initial message payload of a request is a slice of the socket input
  buffer.
* Hence, when the request-on-conn suspends, the contents of the initial
  message payload may change.
* Moving the sending of the un/supported code into swarm changed the behaviour of
  request handlers, in that the handler might be called after the
  request-on-conn has been suspended and resumed. Thus, the contents of the
  initial payload buffer might have changed. = data corruption.

This change also has the nice side-effect of moving a lot of request
handling boilerplate into swarm.

Fixes #227.

(Note that this is a breaking change to an as yet unreleased new feature of v4.6.0.)

nemanja-boric-sociomantic commented on src/swarm/neo/node/ConnectionHandler.d:
```diff
@@ -357,6 +375,43 @@ class ConnectionHandler : IConnectionHandler
+
+    ***************************************************************************/
+
+    private void[] emplace_buf;
```
Shouldn't this be on the `RequestOnConn` level, given that the single connection handles many requests?

gavin-norman-sociomantic replied:
Haha, probably yes.

gavin-norman-sociomantic replied:
That'd be a big advantage of stack allocation, actually.

nemanja-boric-sociomantic replied:
Could we plug in swarms `AcquiredBuffers` here? So that the `handleRequest` gets a buffer and recycles it?

gavin-norman-sociomantic replied:
Hmm... In theory, yes. The code in swarm doesn't currently know anything about any kind of shared resources, though.

nemanja-boric-sociomantic replied:
No, I mean just to have reusable buffers, nothing else.

gavin-norman-sociomantic replied:
What would be the advantage?

nemanja-boric-sociomantic replied:
I wrote that comment before I've heard it's easy enough to add it as a member of the `RequestOnConn` :-)

gavin-norman-sociomantic replied:
Updated.

Commit:
Modify per-version request handler support to use ClassInfo, not handler fn

The impetus for this change is to fix the issue discovered in
sociomantic-tsunami/dhtproto#93:
* The initial message payload of a request is a slice of the socket input
  buffer.
* Hence, when the request-on-conn suspends, the contents of the initial
  message payload may change.
* Moving the sending of the un/supported code into changed the behaviour of
  request handlers, in that the handler might be called after the
  request-on-conn has been suspended and resumed. Thus, the contents of the
  initial payload buffer might have changed. = data corruption.

This change also has the nice side-effect of moving a lot of request
handling boilerplate into swarm.

Fixes #227.

nemanja-boric-sociomantic commented on src/swarm/neo/node/ConnectionHandler.d:
```diff
@@ -471,9 +523,26 @@ class ConnectionHandler : IConnectionHandler
+                                {
+                                    auto rq_handler =
+                                        this.emplace!(IRequestHandler)
+                                        (connection.emplace_buf, rq.class_info);
```
Ah, cool :+1:. I'll never be able to intuitively figure out the fact that the `connection` is actually `RequestOnConn` :-(

gavin-norman-sociomantic replied:
Yeah, we should probably go through and make the doc / variable names consistent, at some point.

gavin-norman-sociomantic commented:
(Blocked until tested in dhtproto.)

gavin-norman-sociomantic commented:
Three requests in dhtproto converted; ok so far: https://github.com/sociomantic-tsunami/dhtproto/pull/94

gavin-norman-sociomantic commented on src/swarm/node/model/NeoNode.d:
```diff
@@ -894,6 +894,24 @@ public class NodeBase ( ConnHandler : ISwarmConnectionHandler ) : INodeBase
+
+    ***************************************************************************/
+
+    abstract protected void requestAcquirer ( Object shared_resources,
```
Oof, wait, this is a breaking change.

nemanja-boric-sociomantic replied:
I assumed we're targeting v5.x.x :-)

gavin-norman-sociomantic replied:
Nope. v4.6.0.

gavin-norman-sociomantic replied:
Updated to non-breaking change.

Commit:
Store shared resources in node

Previously, this was passed into the ctor via the Options struct, but was
only passed on to the connection setup params, not stored in the node
itself. Storing it in the node simplifies future refactoring.

Commit:
Adapt test suite to use new request handler map setup

gavin-norman-sociomantic commented:
Updated with a few simplifications.

nemanja-boric-sociomantic commented on src/swarm/node/model/NeoNode.d:
```diff
@@ -618,7 +618,7 @@ public abstract class INodeBase : INode, INodeInfo
 *******************************************************************************/
 
-public class NodeBase ( ConnHandler : ISwarmConnectionHandler ) : INodeBase
+public abstract class NodeBase ( ConnHandler : ISwarmConnectionHandler ) : INodeBase
```
`abstract`

nemanja-boric-sociomantic replied:
This is still abstract (it shouldn't be).

gavin-norman-sociomantic replied:
Updated.

nemanja-boric-sociomantic commented on integrationtest/neo/node/Node.d:
```diff
@@ -96,6 +96,24 @@ public class Node : NodeBase!(ConnHandler)
+    override protected void getResourceAcquirer (
+        void delegate ( Object resource_acquirer ) handle_request_dg )
+    {
+        handle_request_dg(this.shared_resources);
```
Can't the node base class implement this?

gavin-norman-sociomantic replied:
In the case of this dumb test node, it could. Real node implementations need to do something quite different here, though.

gavin-norman-sociomantic commented:
Tests pass in dhtproto, with this change. (See https://github.com/sociomantic-tsunami/dhtproto/pull/94.) Unblocked.
